contracts/gauge-borrowing/VaultPriceOracle.sol

```diff
@@ -0,0 +1,188 @@
+import "../interfaces/Compound/ILiquidityGauge.sol";
+import "../interfaces/Compound/IRibbonVault.sol";
+import "hardhat/console.sol";
+/**
+ * @title VaultPriceOracle
+ * @notice Returns prices from Chainlink.
+ * @dev Implements `PriceOracle`.
+ * @author David Lucid <david@rari.capital> (https://github.com/davidlucid)
+ */
+contract VaultPriceOracle is IPriceOracle, IBasePriceOracle {
+    using SafeMathUpgradeable for uint256;
+
+    /**
+     * @notice Maps ERC20 token addresses to ETH-based Chainlink price feed contracts.
+     */
+    mapping(address => IAggregatorV3Interface) public priceFeeds;
+
+    /**
+     * @notice Maps ERC20 token addresses to enums indicating the base currency of the feed.
+     */
+    mapping(address => FeedBaseCurrency) public feedBaseCurrencies;
+
+    /**
+     * @notice Enum indicating the base currency of a Chainlink price feed.
+     */
+    enum FeedBaseCurrency {
+        ETH,
+        USD,
+        BTC
+    }
+
+    /**
+     * @notice Chainlink ETH/USD price feed contracts.
+     */
+    IAggregatorV3Interface public constant ETH_USD_PRICE_FEED = IAggregatorV3Interface(0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419);
+
+    /**
+     * @notice Chainlink BTC/ETH price feed contracts.
+     */
+    IAggregatorV3Interface public constant BTC_ETH_PRICE_FEED = IAggregatorV3Interface(0xdeb288F737066589598e9214E782fa5A8eD689e8);
+
+    /**
+     * @dev The administrator of this `MasterPriceOracle`.
+     */
+    address public admin;
+
+    /**
+     * @dev Controls if `admin` can overwrite existing assignments of oracles to underlying tokens.
+     */
+    bool public canAdminOverwrite;
+
+    /**
+     * @dev Constructor to set admin and canAdminOverwrite.
+     */
+    constructor (address _admin, bool _canAdminOverwrite) public {
+        admin = _admin;
+        canAdminOverwrite = _canAdminOverwrite;
+    }
+
+    /**
+     * @dev Changes the admin and emits an event.
+     */
+    function changeAdmin(address newAdmin) external onlyAdmin {
+        address oldAdmin = admin;
+        admin = newAdmin;
+        emit NewAdmin(oldAdmin, newAdmin);
+    }
+
+    /**
+     * @dev Event emitted when `admin` is changed.
+     */
+    event NewAdmin(address oldAdmin, address newAdmin);
+
+    /**
+     * @dev Modifier that checks if `msg.sender == admin`.
+     */
+    modifier onlyAdmin {
+        require(msg.sender == admin, "Sender is not the admin.");
+        _;
+    }
+
+    /**
+     * @dev Admin-only function to set price feeds.
+     * @param underlyings Underlying token addresses for which to set price feeds. (gauge token)
+     * @param feeds The Chainlink price feed contract addresses for each of `underlyings`. (underlying of gauge's vault token)
+     * @param baseCurrency The currency in which `feeds` are based.
+     */
+    function setPriceFeeds(address[] memory underlyings, IAggregatorV3Interface[] memory feeds, FeedBaseCurrency baseCurrency) external onlyAdmin {
```

Comment:
We can use calldata over memory since we aren't mutating to save a bit of gas.

https://twitter.com/0xkarmacoma/status/1487219340277534725